### DISCUSSION TITLE
KPH [The Hardening] [Project Update]

### DISCUSSION THREAD

**Discussion by jxy-s** [winsiderss/systeminformer#773] [Questions / Answers]
We're updating the kernel driver functionality (KPH). The focus of these changes will include hardening KPH with extra protections and provide more information to be displayed in PH.

Updates and announcements will be posted on the [issue page](https://github.com/processhacker/processhacker/issues/774).

If you have questions or feature requests post them here.

**Comment by dmex** [comment]
Windows 7 and 8 support will be dropped in the next update.

The other KPH updates include:

* Nested Job object support.
* Improved file names for console device handles.
* Improved thread stack walking.
* Improved service logging.
* Improved firewall/network logging.
* Performance improvements.

Some might not be included in the next update but likely included in later updates.

**Reply by Biswa96** [reply]
If Windows 7 support will be dropped, I would suggest to make a tag/release for that only. We're talking about [100M PCs][1] 😜 

[1]: https://www.theverge.com/2021/1/6/22217052/microsoft-windows-7-109-million-pcs-usage-stats-analytics

**Reply by dmex** [reply]
@Biswa96 

> We're talking about 100M

Telemetry shows Windows 7 is only 14% of users and falling:

![image](https://user-images.githubusercontent.com/1306177/104493417-1e7ba680-5629-11eb-8bc6-6689435d5f5c.png)

> suggest to make a tag/release for that only

Windows 7/8 might still be supported client-only but driver support won't be continued.  If the support for Windows 7 and 8 is dropped then it'll  be the same situation as XP and Vista with downloads added to the Legacy Operating Systems:
https://processhacker.sourceforge.io/downloads.php

**Comment by MagicAndre1981** [comment]
I till use Windows 8.1 as my main system. It is rock stable and not so buggy as hell as Windows 10.

**Reply by trparky** [reply]
Windows 10 is no less buggy than any other version of Windows.

**Comment by diversenok** [comment]
I wonder: what is the point of doing a domination check on the protection level if PH cannot pass it? You can merely say that it removes the functionality, yielding Process Hacker useless against malware that somehow toggles process protection on itself.  So, perhaps, you're planning to let PH run as a protected process, and KPH Hardening is one of the steps to make sure nobody can misuse this functionality?

**Reply by jxy-s** [reply]
For the time being, yes. This is to mitigate abuse of PH to disable protected processes. Eventually by some means we plan to give back the functionality to process hacker.

**Comment by MPeti1** [comment]
I've recently read on an [other project](https://github.com/eiz/SynchronousAudioRouter/issues/86) that MS is again making it harder to get your drivers signed so that Windows will load it without the user first disabling certain system protections. After that happens, how will you release new updates to the driver?

**Comment by DavidXanatos** [comment]
Please do not drop windows 7 support, what would be the point of that anyways? 
I can't imagine keeping these support would be so much extra maintenance afford.


btw: what are you using for these telemetry?

**Comment by DavidXanatos** [comment]
I had a look into the recent changes and it seams you have removed the ability to disable client process signature verification all together.
May I suggest to give the user back the option to use the driver in an insecure way if he so chooses by disabling all safety checks except requiring admin or debug privileges by the calling process, if the system is started in test signing mode.

In such scenarios the user could just run an unsigned driver with all safety checks disabled anyways, but that would be  work and afford to do, so adding this functionality to the original driver would be very useful for many people.
You would only need to query 
```
    SYSTEM_CODEINTEGRITY_INFORMATION sci = {sizeof(SYSTEM_CODEINTEGRITY_INFORMATION)};
	if(NT_SUCCESS(ZwQuerySystemInformation(/*SystemCodeIntegrityInformation*/ 103, &sci, sizeof(sci), NULL)))
```
and act accordingly.

Also it would be a nice improvement to add a KernelDbgView functionality to the driver and client process as I did for TE, see: https://github.com/DavidXanatos/TaskExplorer/blob/master/ProcessHacker/XProcessHacker/log.c

**Reply by jxy-s** [reply]
> May I suggest to give the user back the option to use the driver in an insecure way if he so chooses by disabling all safety checks except requiring admin or debug privileges by the calling process, if the system is started in test signing mode.

Releasing a signed driver with the ability to weaken system security boundaries is not something I'm comfortable with. Even with measures in place attempting to mitigate risk. That said, the usefulness of KPH APIs from other clients is not lost on me. At that rate, I think it best for a user to compile their own version of KPH with these mitigations disabled. We could implement compile-time option for that make it easier for a user to compile with those mitigations off.

> Also it would be a nice improvement to add a KernelDbgView functionality.

This is on my todo list already 😄. There are a other hardening changes prior to adding new functionality for KPH that are queuing up. Unfortunately we've got some roadblocks on Microsoft's end and we're holding off on PRs until those blockers are sorted.

**Reply by DavidXanatos** [reply]
I see... 

btw just out of curiosity what troubles is MSFT doing? What don't they like about the driver?

**Reply by jxy-s** [reply]
Microsoft's signing process has been the blocker for us.

**Reply by DavidXanatos** [reply]
So they outright refuse to sign your driver? Only the PH driver or any driver?
Why is the review process so thorough? For my sandboxie we just upload the binary and the signed driver pops out, I don't think a human even takes a look on it.
Or do they have some anti malware scanners triggered when they see KPH code?

**Reply by jxy-s** [reply]
Yeah it is usually that simple, I agree 😄.

I don't have more information on that end of things - @dmex has been handling that. We've got some resources to pull on to get things moving. And I'm here to support/help @dmex in whatever he needs to move that ball forward.

**Reply by DavidXanatos** [reply]
I'm worried that MSFT may make more trouble here, I have noticed that in w11 they finally blacklisted all the old leaked signing certificates. And they are pushing secure boot and TPM as mandatory, so a driver like KPH may be really not what they would want when pushing the iOS-ification of windows forwards in big steps towards a locked down non user controlled system :(
I'm curious to learn more details about the situation as for my TaskExplorer fork of PH I'll need to sign the driver properly for w11 as well, and I must avoid "burning" the EV cert I'm using with sandboxie.

**Reply by dmex** [reply]
@DavidXanatos 

> So they outright refuse to sign your driver?

The signing process fails each time without any error messages and Microsoft claimed "this surpasses our support"... They've just fucked me around endlessly until the certificates expire.

![image](https://user-images.githubusercontent.com/1306177/127983792-1d810242-2043-47b5-bc78-848e838dfe1e.png)


> Or do they have some anti malware scanners triggered when they see KPH code?

This seems to be the case when submitting the driver. The exact same issue happened when submitting to Microsoft Winget:

![image](https://user-images.githubusercontent.com/1306177/127981554-95cb4985-be54-4f3a-abf7-135abeb8d467.png)

I tired emailing him but never got a response about this behavior. You can also see how many times the package failed for unexplained reasons and that exact same problem happens when submitting the driver:

https://github.com/microsoft/winget-pkgs/pull/373

> a driver like KPH may be really not what they would want

Microsoft Process Explorer has the same functionality so they don't have standing to block competitors then go and include the exact same features in their own software.

Microsoft has been secretly adding more powerful features than Process Hacker via their SAC product - SAC has no security whatsoever by design - they're clearly targeting the project not because of any actual technical issues but rather because we're more popular than their products, so they're using the same (illegal and anti-competitive) tactics they used against Netscape Navigator to eliminate competition but also labeling the project malicious in an attempt to mislead the competition regulators.

There's been some discussion about this in #454

> I have noticed that in w11 they finally blacklisted all the old leaked signing certificates. And they are pushing secure boot and TPM as mandatory

The existing drivers are compatible with Win11 and haven't been blocked by Microsoft yet... The large majority of changes by Microsoft are limited to restricting the Windows API with signature checks that block competitors software (e.g. CreateWindowInBand, NtQuerySystemInformation, NtQueryInformationProcess to name a few) rather than directly targeting the drivers themselves.

The signature checks added to those functions and classes only block third-parties and this includes signed binaries. We won't be able to implement the same functionality as Task Manager and Process Explorer because of those Microsoft-only signature checks even after we sort out the submission issue.

Always-on-top, Auto-elevation, DPS statistics, Default taskmgr application preferences (Microsoft hardcoded taskmgr.exe blocking competitors), GPU statistics (deliberately broken on Win10 and Win11 recently) and the DirectUI framework are some examples of features that I want to implement and are currently implemented by Task Manager but are Microsoft-only signature restricted while newer more advanced security like PPL that we desperately need are also Microsoft-only signature restricted.

The only certificate allowed to use these and other functionality is now limited to Microsoft Windows certificates - the same certificates used with Task Manager and Process Explorer - while SAC has even more powerful functionality than anything else (including Process Hacker) with absolutely no security whatsoever.

> I'm worried that MSFT may make more trouble here

I've been complaining to Microsoft employees for years about this stuff but the attacks keep getting worse and I've since started demanding our competition regulator prosecute the company after they labeled the project malicious last year... Microsoft claims to love open source and be more transparent these days but the bullshit they're doing with SAC, taskmgr and procxp while attacking competitors and trying to limit competition and kill off the project is insane.

> KPH may be really not what they would want when pushing the iOS-ification of windows forwards in big steps towards a locked down non user controlled system

Why don't they just come out and say this? Why go the asshole route with illegally hard-coding taskmgr.exe/procxp.exe and limiting functionality to those products while blocking competitors? I was around during the 90's and they killed Netscape with this exact same behavior by changing APIs and blocking Netscape from those same APIs.

Windows owns the market for the simple reason it's not some locked down garbage controlled system so they need to start communicating these changes if they intend to kill off third party task managers or instead doing something about the numerous complaints and issues that I have complained about or they'll end up getting prosecuted and charged by regulators again just like last time when they did this exact same bullshit with Netscape.

**Reply by DavidXanatos** [reply]
hmm not good..
btw what does SAC stand for?

**Reply by DavidXanatos** [reply]
PS: can you still submit other derives and have them successfully signed?

**Reply by jxy-s** [reply]
@DavidXanatos, @dmex and I spoke offline on the driver signing topic - he has gotten me up to speed and I will be heading this effort moving forward.

>I'm curious to learn more details about the situation as for my TaskExplorer fork of PH I'll need to sign the driver properly for w11 as well, and I must avoid "burning" the EV cert I'm using with sandboxie.

I want to directly address your comment - at this time I don't have evidence to justify this concern.

> can you still submit other derives and have them successfully signed?

I'll will work to answer this question once I have the facility to do so. We have some organizational work to do in sysdev (Hardware Partner Portal) before I can begin testing myself.

Thanks for your interest and participation here, I'll update the thread when I have more information 👍.

**Reply by DavidXanatos** [reply]
Could someone please tell me what SAC stands for?
Probably in this context not "Semi Annual Channel" lol but as msft used this abbreviation for that it hard to find anything else with google for this "SAC"

**Reply by dmex** [reply]
> Could someone please tell me what SAC stands for?

Special Administration Console (SAC)

`dism.exe /online /get-capabilityinfo /CapabilityName:Windows.Desktop.EMS-SAC.Tools~~~~0.0.1.0`
and
`dism.exe /online /add-capability /CapabilityName:Windows.Desktop.EMS-SAC.Tools~~~~0.0.1.0`

![image](https://user-images.githubusercontent.com/1306177/129475904-66f6f0f6-0fb6-4417-905d-994b41a49ab1.png)


The SAC toolset is a boot-start kernel driver copied into `C:\Windows\system32\DRIVERS\sacdrv.sys` which is also installed by default on every version of Windows Server... The documentation states it's only available by RS232 and USB ports however hardware vendors include a [SPCR (Console Redirection Table)](https://docs.microsoft.com/en-us/windows-hardware/drivers/serports/serial-port-console-redirection-table) in the ACPI table and redirect this port over the network.

For example this screenshot from a vendor showing the port available over the internet via SPCR:

![image](https://user-images.githubusercontent.com/1306177/129474438-940eac57-fe77-4cce-8ecd-b28767fd0eb9.png)

There is zero authentication required to enumerate and terminate processes... Input T and press enter to show the process list:

![image](https://user-images.githubusercontent.com/1306177/129474475-e84595c5-0988-4cba-bbb7-afb7ed8dce81.png)


Input K to terminate processes:

![image](https://user-images.githubusercontent.com/1306177/129474575-26e9bc81-d030-44f7-990e-5a41f6feacf1.png)

This command supports terminating PPL processes and bypasses all security products including security products such as Windows Defender.

Not only third parties make SAC internet accessible but also Intel AMT and AMD's DASH both have `SOL/LMS drivers` redirecting Windows SAC over the network using their own protocols. It's much more powerful than Process Hacker since SAC doesn't require authentication and doesn't have any security checks whatsoever when terminating processes and it's network accessible... Process Hacker doesn't have any remote management features/network management capability, requires an interactive user desktop session with an _authenticated_ administrator via a High integrity process and requires debugger privileges.

SAC is also enabled by default for all Windows Azure VMs accessible via the Console:
![image](https://user-images.githubusercontent.com/1306177/129475496-f4b5df67-0bb4-4be0-b213-470c8e6b0fcc.png)

The entire time Microsoft has been targeting Process Hacker for our termination feature they've been making SAC termination more powerful than PH, removing authentication requirements/security checks when terminating processes, making SAC internet accessible, allowing vendors to silently install the feature via the SPCR in the ACPI table and install SOL/LMS drivers.

All the while Microsoft release hundreds of signatures targeting our binaries as malicious (#454), block the website and download URLs as malicious #904 and allow employees to spread FUD (#620)... and completely refuse to discuss anything like we don't exist.

**Reply by DavidXanatos** [reply]
Interesting driver,
is the process termination feature of PH the only thing MSFT has a problem with?
I mean if its the only thing they don't like, may be its worth moving that feature into a separate tool or a plugin with an own unsigned driver?
Also changing the name would be an option, if than all the problems can be avoided.

**Reply by dmex** [reply]
> the only thing MSFT has a problem with

MS refused to discuss anything and have ignored every email so who knows what their problem is.

> if its the only thing they don't like

It's not the only thing.... There are recent changes to APIs that block and limit features when the caller isn't taskmgr.exe.

Either way this discussion is offtopic from the KPH updates.

**Comment by AzAgarampur** [comment]
Can ProcessHacker.exe be signed as PPL via a certificate, and then KProcessHacker use `PsGetProcessProtection` in `IRP_MJ_CREATE` to make sure caller is protected? This way no other random caller can use the driver.

**Reply by roblabla** [reply]
> Can ProcessHacker.exe be signed as PPL via a certificate

AFAICT, there are only two kinds of processes that can elevate to PPL:

- Processes signed by microsoft themselves, only for their own binaries
- Processes signed by a certificate contained within an Early-Launch Anti-Malware driver (ELAM), see [Protecting Anti-Malware Services](https://docs.microsoft.com/en-us/windows/win32/services/protecting-anti-malware-services-#requirements).

The former is obviously impossible for ProcessHacker, and the later requires an ELAM driver, which comes with tons of strings attached.

**Comment by jxy-s** [comment]
I want to provide an update here. 

I have been working Microsoft to ensure the next release of System Informer is shipped with a driver that complies with their requests.

I have updated the driver with enhanced mitigations. I have also added some of the suggested functionality here. This is not yet fully complete. It has not been merged into the main repo yet. I will be adding more features to the driver over time.

I will be providing a signed driver to the HVCI team prior to shipping to ensure they are on-board with the changes.

I had previously wanted to target only Windows 10+ with the new driver. I'm pleased to say I will be providing _limited_ support for Windows 7 SP1 through Windows 8.1. However long-term support for those legacy OSes is not happening. Meaning the driver will function on those OSes but is _not_ fully supported/maintained.

**Reply by DavidXanatos** [reply]
Sooo... if I may speculate, you are going to get an ELAM driver for PH?

**Reply by jxy-s** [reply]
No plans for ELAM right now.

**Reply by DavidXanatos** [reply]
oh... so also no running as PPL? that's a pity, would be quite cool :D

**Reply by jxy-s** [reply]
Yeah it would be. Maybe in the future. But it's not in scope for now.

**Reply by MPeti1** [reply]
I'm a user of Process Hacker, and have been following the topic for a longer time, out of curiosity of how (if ever) my favourite task manager's devs resolve this driver signing problem, despite problems on Microsoft's side.

I see that the discussion, and actually Process Hacker's repository was moved to a new organization and also renamed. It's a bit confusing, and to be honest I'm uneasy about this, because it is not explained what is going on.
I have also seen #1286.

Could you or an other (formerly known) project member please clarify what has happened?
Some more specific questions of mine: Who are behind Winsider Seminars & Solutions, Inc.? Does Microsoft have influence on the future of the project? Will the former, known maintainers (wj32, dmex and you) still be part of PH (SI?) with the same level of access?

I don't intend to be intrusive, but you may understand that the reason I (and possibly other users of PH) would like to know this is because as you know, the kernel driver grants very deep access to our systems, but even when not using it, it's good to know who are behind software we use. Some of the questions may be weird, but I think not unreasonable.

Also, you said in the referenced issue that currently it is expected that master is unstable. Does that also mean that it is wiser to not allow updates to PH until that changes?

I'm aware that you have a lot to do, please know that I don't expect immediate response, it's ok if it will have to wait some time.
In the meantime I hope the best, and please keep up the good work!

**Reply by AltF5** [reply]
Only the person that helped revise Windows Internals books into edition 6 & 7. Alex is awesome and one of the best Security Researchers and OS devs, in the likes of a hybrid between James Forshaw and Mark Russinovich

https://twitter.com/aionescu/status/1536494552105766912

It’s in great hands.

I don’t want to @ mention those persons here since they consider doing so as spam to @ the code maintainers directly If they haven’t commented. But they have been involved here for years prior.

I highly recommend you follow him on twitter and also the other person he mentioned in that tweet: yarden 

Twitter is good for the latest and greatest Sec news. Dmex (Stephen) doesn’t have a twitter  but there is @ ProcessHacker which may or may not be run by him… there is also a newly created @ SystemInformer probably by Alex. If you don’t already follow jxy and diversenok on twitter then I’d recommend them as well for the latest news 🗞 when announced 

Regarding wj32 he’s unfortunately no longer involved at least since 2017/2018 or so and dmex is his successor.

Regarding what happened, here’s a humorous, but confirmed true, take on the “order of operations”
https://twitter.com/gabriellandau/status/1536730125886267393

Regarding MS’ involvement: only as involved as getting a WHQL-signed driver. Things are already less malware like by getting away from the “hacker” perspective and renaming a lot, but to avoid having people enable TestSigning mode a signed driver will make things more legit. Note it wont be an ELAM-signed driver (Early Launch Anti Malware) which could make PH PPL (Protected Process, only terminate or or accessible from other PPLs).

Disclaimer: I’m a power user, and not associated with the core maintainers, other than DM communication with them on twitter (for those that have it)

**Reply by DavidXanatos** [reply]
IMHO this is a very promising development, I'm quite optimistic for a bright future of this project. 
Just the name sounds a bit boring, but then finding a new unique name for a task manager is not that easy, all the good once were taken decades ago. And to clean up with the bad hacker image it probably needed to sound as harmless as possible. System Spy sounds suspicious, already, anything with hacker is a no go, etc....
But it might have been a better choice to call it System Insider, Informer sounds so very STASI like LOL. On the other hand System informer really does what it says it does, it snitches on the system to the user LOL.

**Reply by jxy-s** [reply]
@dmex and I will continue to maintain and run the project. We're not going anywhere :).

**Reply by AltF5** [reply]
Trust me “Insider” is not less evoking than “hacker”. Arguably more red flagy 🚩

**Reply by MPeti1** [reply]
Thank you all for the responses, this sounds pretty good! keep up the good work! :)

**Comment by jxy-s** [comment]
I've posted a pull request for this work (https://github.com/winsiderss/systeminformer/pull/1320). Not all the improvements we've wanted to make have been implemented. Those will come in future work.

**Comment by jxy-s** [comment]
> I had previously wanted to target only Windows 10+ with the new driver. I'm pleased to say I will be providing _limited_ support for Windows 7 SP1 through Windows 8.1. However long-term support for those legacy OSes is not happening. Meaning the driver will function on those OSes but is _not_ fully supported/maintained.

Turns out that signing a driver to support Win7 is going to be logistically impossible for us right now. Due to these signing restrictions and telemetry indicating that a miniscule number of users are on legacy OSes or use 32bit. We are only going to release for x64 and AMD64 Win10 and above.

That said, I'm going to keep the code which supports the legacy OSes in the repo for the time being. So those using or doing research on those OSes may build and use the driver themselves.

For reference:
https://techcommunity.microsoft.com/t5/windows-hardware-certification/changes-to-driver-publication-for-windows-7-sp1-windows-server/ba-p/2459992
https://techcommunity.microsoft.com/t5/windows-hardware-certification/changes-to-driver-signing-for-windows-7-windows-server-2008-r2/ba-p/2460039
This new process is only for HCK drivers, which we are not, and will end in January anyway.

**Reply by DavidXanatos** [reply]
I'll gladly help you figuring out how to get the MSFT portal to generate a working windows 7 signature, after we understood how to do that it was trivial. 
Sandboxie hence comes with full windows 7 support and will do so for quite a few more years to come.


PS: what sort of telemetry do you employ?

**Reply by jxy-s** [reply]
@ionescu007 working on the signing for me. Please provide him with the information you have.

**Reply by jxy-s** [reply]
Telemetry is above in the thread. @dmex can provide more details.

**Reply by DavidXanatos** [reply]
@jxy-s  @ionescu007 Anything new on this subject?

Without knowing what you are failing at, here is our experience:
Our problem with windows 7 was that we assumed the files we need to upload to MSFT for signing are the exact same we ran the (as we learned later unneeded) WHQL testing, that is the test signed files.
Which was a mistake, it is needed to strip any signature from the files, as MSFT only appends their own in addition to what is there.
And windows 7 only checks the first found signature, that was the test signing one which does not work in production.

Uploading completely unsigned sys files for attestation testing in the MSFT portal did the trick, the returned sys files had only one MSFT signature and were recognized by windows 7 just fine.

**Reply by dmex** [reply]
@DavidXanatos We've been told that as of May the root certs expired in the ci.dll and even if you can sign drivers for Windows 7 they don't work?

**Reply by DavidXanatos** [reply]
@dmex the SbieDrv.sys which I release a new build about every week or two (only signed by MSFT's portal) seam to work just fine on windows 7 since may no one complained about any driver issues, except if they were missing the patches which bring sha256 support. So perhaps windows 7 just does not care about the expiration date.

**Reply by andry81** [reply]
> telemetry indicating that a miniscule number of users are on legacy OSes

How did you find out this if telemetry is the first thing which does turn off on legacy OSes including all services to run faster under VM which is not a user at all?

**Reply by dmex** [reply]
> How did you find out this if telemetry is the first thing which does turn off on legacy OSes

We get statistics (telemetry) from Sourceforge and Github when you download binaries. We do not collect anything from the operating system.

**Reply by andry81** [reply]
> > How did you find out this if telemetry is the first thing which does turn off on legacy OSes
> 
> We get statistics (telemetry) from Sourceforge and Github when you download binaries. We do not collect anything from the operating system.

This is incomplete or distorted statistic. You can get it the same way, for example, in the Steam, where the user just have no choice to keep the old OS and Video Card, because people, you know, play games. In other words this statistic is not related to SF or GH.

**Reply by trparky** [reply]
First off, Windows 7 is how old? And not only that but it's been an unsupported OS nearly three years. Time to put it to bed already. I support DMEX's decision to pull support for Windows 7 especially if it'll cause undue stress on him to continue supporting it.

**Reply by DavidXanatos** [reply]
@trparky what are you talking about? Windows 7 still receives ESU Support and is in use world wide, MSFT is even considering to extend this support even longer. IMHO software should support an OS at least 3 years past the very last update that OS received. So for windows 7 there are at least 3 more years to go.

**Reply by trparky** [reply]
> @trparky what are you talking about? Windows 7 still receives ESU Support and is in use world wide,

Yes, at an ever-increasing cost; yes. Every year that goes by Microsoft charges you more for that ESU support. At what point do you finally say... Nope. It's cheaper to upgrade than to keep paying out for additional support.

And by support, I mean mainstream support. The average person on the street doesn't have that option, it's either update or not get updates.

**Reply by DavidXanatos** [reply]
> > @trparky what are you talking about? Windows 7 still receives ESU Support and is in use world wide,
> 
> Yes, at an ever-increasing cost; yes. Every year that goes by Microsoft charges you more for that ESU support. At what point do you finally say... Nope. It's cheaper to upgrade than to keep paying out for additional support.
> 
> And by support, I mean mainstream support. The average person on the street doesn't have that option, it's either update or not get updates.

Well do you know how many 10's of 1000's of € various specialized lab equipment costs? We have even windows XP machines running some of them on real hardware because we are not going to scrap a 100k instrument only because of missing OS updates.

And about your "average person" they eider don't care for updates only for working software or use some ESU bypass hack.

Besides, I don't believe windows 7 support is any meaningful amount of extra work, at least that's what I learned form over 2 years of Sandboxie development.

**Reply by trparky** [reply]
> Well do you know how many 10's of 1000's of € various specialized lab equipment costs? We have even windows XP machines running some of them on real hardware because we are not going to scrap a 100k instrument only because of missing OS updates.

That's when you firewall the hell out of whatever part of the network that they're on, even VLAN them off or put them on different subnet, and leave them be. Don't even think about letting them out onto the public Internet.

> Besides, I don't believe windows 7 support is any meaningful amount of extra work, at least that's what I learned form over 2 years of Sandboxie development.

Sure, that's because Sandboxie is an older program that was written back in the day when Windows 7 was new. If you were to decide to do a new, from the ground up, version of Sandboxie, I imagine you're going to be in for a headache and a half.

**Reply by DavidXanatos** [reply]
> > Besides, I don't believe windows 7 support is any meaningful amount of extra work, at least that's what I learned form over 2 years of Sandboxie development.
> 
> Sure, that's because Sandboxie is an older program that was written back in the day when Windows 7 was new. If you were to decide to do a new, from the ground up, version of Sandboxie, I imagine you're going to be in for a headache and a half.

Guess what Process hacker is also quite old, so just leave the win 7 code in and keep supporting it till kingdom comes.

**Reply by DavidXanatos** [reply]
@jxy-s could you provide the KPH_DYN_CONFIGURATION for windows 7 x64 and 8.1 x64 please.

**Reply by Masamune3210** [reply]
> > > Besides, I don't believe windows 7 support is any meaningful amount of extra work, at least that's what I learned form over 2 years of Sandboxie development.
> > 
> > Sure, that's because Sandboxie is an older program that was written back in the day when Windows 7 was new. If you were to decide to do a new, from the ground up, version of Sandboxie, I imagine you're going to be in for a headache and a half.
> 
> Guess what Process hacker is also quite old, so just leave the win 7 code in and keep supporting it till kingdom comes.

Time is not a infinite commodity and the longer code sits around, the more time it takes and the harder it gets to work around. If you want to keep support, nobody is stopping you from just not updating to the new version if one comes out that drops support. Or, even better, if you see the value in keeping support, the project is open source, feel free to contribute help to keep the support

**Reply by DavidXanatos** [reply]
@Masamune3210 the header of the relevant fiel says its auto generated, but o cant find the tool online so the devs have a tool that makes extracting the relevant offsets easy, hence my request for them to provide the offsets for windows 7 and 8.1. If they prefer to make the tool public that would be even better.

**Reply by dmex** [reply]
> @jxy-s could you provide the KPH_DYN_CONFIGURATION for windows 7 x64 and 8.1 x64 please.

It's been here since the first release?
https://github.com/winsiderss/systeminformer/blob/018b5c11bd0dcaab6b6e5c140787fafc97ba1b4f/kphlib/kphdyn.xml#L244-L321

> the header of the relevant fiel says its auto generated, but o cant find the tool online so the devs have a tool that makes extracting the relevant offsets easy

There is no automatic extraction. All offsets are copied manually from windbg into that XML file then we execute `CustomBuildTool.exe -dyndata` to load the XML and generate the header:

https://github.com/winsiderss/systeminformer/blob/018b5c11bd0dcaab6b6e5c140787fafc97ba1b4f/tools/CustomBuildTool/Program.cs#L38-L41

> If they prefer to make the tool public that would be even better.

Everything is public?

**Reply by DavidXanatos** [reply]
Thanks, somehow id not find that, thanks for pointing out, LOL

**Reply by i486** [reply]
@DavidXanatos,

I'm curious if your method could apply to Nvidia drivers. They're still rolling out new Windows 7 drivers without signatures. Us Windows 7 users are left with either using TESTSIGNING or signing them ourselves with custom certs. Maybe it's time to give Nvidia a heads-up about this, haha.

Apologies for going off-topic.
